Cached sources "couldn't be obtained" when bzip2 is missing

**EDIT 2**: See https://github.com/ocaml/opam/issues/3497#issuecomment-410368271 for the real reason for this error.  In short, I didn't have `bzip2` installed but the error message from `opam` was misleading.

A fresh opam 2 (2.0.0-rc4) install on CentOS 7 with bubblewrap 0.3.0, using the 64bit Linux build of opam available on github I get the following error from a freshly created 4.06.1 switch:
```
$ opam install alcotest                                                                                                                                                                                                                                                                    
The following actions will be performed:
  ∗ install ocamlbuild 0.12.0     [required by cmdliner, astring, fmt]
  ∗ install dune       1.0.1      [required by jbuilder]
  ∗ install conf-m4    1          [required by ocamlfind]
  ∗ install uchar      0.0.2      [required by fmt]
  ∗ install jbuilder   transition [required by alcotest]
  ∗ install ocamlfind  1.8.0      [required by cmdliner, astring, fmt]
  ∗ install result     1.3        [required by alcotest]
  ∗ install base-bytes base       [required by astring]
  ∗ install topkg      0.9.1      [required by cmdliner, astring, fmt]
  ∗ install cmdliner   1.0.2      [required by alcotest]
  ∗ install astring    0.8.3      [required by alcotest]
  ∗ install fmt        0.8.5      [required by alcotest]
  ∗ install alcotest   0.8.3
===== ∗ 13 =====
Do you want to continue? [Y/n] y

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[alcotest.0.8.3] found in cache
[astring.0.8.3] found in cache
[cmdliner.1.0.2] found in cache
[dune.1.0.1] found in cache
[fmt.0.8.5] found in cache
[ocamlbuild.0.12.0] found in cache
[ocamlfind.1.8.0] found in cache
[result.1.3] found in cache
[topkg.0.9.1] found in cache
[uchar.0.0.2] found in cache
[ERROR] The sources of the following couldn't be obtained, aborting:
          - alcotest.0.8.3
          - astring.0.8.3
          - cmdliner.1.0.2
          - dune.1.0.1
          - fmt.0.8.5
          - result.1.3
          - topkg.0.9.1
          - uchar.0.0.2
```

```
$ opam config report
# opam config report
# opam-version      2.0.0 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=centos os-version=7
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              7
# repositories      1 (http), 1 (version-controlled) (default repo at 815ad398)
# pinned            0
# current-switch    4.06.1
```
**EDIT:** I added a note about the version of bubblewrap installed in case it's relevant.

In verbose mode `-vv`, you can see all  commands output. I added a more clear error message for that case, that is shown on verbose mode
```
[ERROR] The sources of the following couldn't be obtained, aborting:
          - alcotest.0.8.3:
            Tar needs bzip2 to extract the archive
```

Thanks for reporting!

It's the default one, for extracting errors:
```
[ERROR] The sources of the following couldn't be obtained, aborting:
          - alcotest.0.8.3: Could not extract archive
```

--- a/src/core/opamSystem.ml
+++ b/src/core/opamSystem.ml
@@ -556,11 +556,29 @@ open OpamProcess.Job.Op
 
 module Tar = struct
 
+  type extract =
+    | Bzip2
+    | Gzip
+    | Lzma
+    | Xz
+
+  let extract_command = function
+    | Bzip2 -> "bzip2"
+    | Gzip -> "gzip"
+    | Lzma -> "lzma"
+    | Xz -> "xz"
+
+  let extract_option = function
+    | Bzip2 -> 'j'
+    | Gzip -> 'z'
+    | Lzma -> 'Y'
+    | Xz -> 'J'
+
   let extensions =
-    [ [ "tar.gz" ; "tgz" ], 'z'
-    ; [ "tar.bz2" ; "tbz" ], 'j'
-    ; [ "tar.xz" ; "txz" ], 'J'
-    ; [ "tar.lzma" ; "tlz" ], 'Y'
+    [ [ "tar.gz" ; "tgz" ], Gzip
+    ; [ "tar.bz2" ; "tbz" ], Bzip2
+    ; [ "tar.xz" ; "txz" ], Xz
+    ; [ "tar.lzma" ; "tlz" ], Lzma
     ]
 
   let guess_type f =
@@ -570,42 +588,53 @@ module Tar = struct
       let c2 = input_char ic in
       close_in ic;
       match c1, c2 with
-      | '\031', '\139' -> Some 'z'
-      | 'B'   , 'Z'    -> Some 'j'
-      | '\xfd', '\x37' -> Some 'J'
-      | '\x5d', '\x00' -> Some 'Y'
+      | '\031', '\139' -> Some Gzip
+      | 'B'   , 'Z'    -> Some Bzip2
+      | '\xfd', '\x37' -> Some Xz
+      | '\x5d', '\x00' -> Some Lzma
       | _              -> None
     with Sys_error _ -> None
 
   let match_ext file ext =
     List.exists (Filename.check_suffix file) ext
 
-  let is_archive f =
-    List.exists
-      (fun suff -> Filename.check_suffix f suff)
-      (List.concat (List.rev_map fst extensions))
-
-  let extract_command file =
-    let command c dir =
-      make_command "tar" [ Printf.sprintf "xf%c" c ; file; "-C" ; dir ]
-    in
+  let get_type file =
     let ext =
       List.fold_left
-        (fun acc (ext, c) -> match acc with
-          | Some f -> Some f
-          | None   ->
-            if match_ext file ext
-            then Some (command c)
-            else None)
+        (fun acc (ext, t) -> match acc with
+           | Some t -> Some t
+           | None   ->
+             if match_ext file ext
+             then Some t
+             else None)
         None
         extensions in
     match ext with
-    | Some f -> Some f
+    | Some t -> Some t
     | None   ->
       match guess_type file with
-      | None   -> None
-      | Some c -> Some (command c)
+      | Some t -> Some t
+      | _ -> None
 
+  let is_archive file =
+    get_type file <> None
+
+  let check_extract file =
+    OpamStd.Option.Op.(
+      get_type file >>= fun typ ->
+      let cmd = extract_command typ in
+      let res = resolve_command cmd <> None in
+      if not res then
+        Some (Printf.sprintf "Tar needs %s to extract the archive" cmd)
+      else None)
+
+  let extract_command file =
+    OpamStd.Option.Op.(
+      get_type file >>| fun typ ->
+      let command c dir =
+        make_command "tar" [ Printf.sprintf "xf%c" c ; file; "-C" ; dir ]
+      in
+      command (extract_option typ))
 end
 
 module Zip = struct
@@ -643,7 +672,11 @@ let extract_job ~dir file =
   | Some cmd ->
     cmd tmp_dir @@> fun r ->
     if not (OpamProcess.is_success r) then
-      Done (Some (Process_error r))
+      if Zip.is_archive file then
+        Done (Some (Process_error r))
+      else match Tar.check_extract file with
+        | None -> Done (Some (Process_error r))
+        | Some s -> Done (Some (Failure s))
     else if try not (Sys.is_directory dir) with Sys_error _ -> false then
       internal_error "Extracting the archive would overwrite %s." dir
     else
@@ -683,9 +716,14 @@ let extract_in_job ~dir file =
   | Some cmd ->
     cmd dir @@> fun r ->
     if not (OpamProcess.is_success r) then
-      Done (Some (Failure
-                    (Printf.sprintf "Failed to extract archive %s: %s" file
-                       (OpamProcess.result_summary r))))
+      if Zip.is_archive file then
+        Done (Some (Process_error r))
+      else match Tar.check_extract file with
+        | None ->
+          Done (Some (Failure
+                        (Printf.sprintf "Failed to extract archive %s: %s" file
+                           (OpamProcess.result_summary r))))
+        | Some s -> Done (Some (Failure s))
     else Done None
 
 let extract_in ~dir file =

--- a/src/repository/opamRepository.ml
+++ b/src/repository/opamRepository.ml
@@ -237,10 +237,8 @@ let pull_tree
     OpamFilename.cleandir local_dirname;
     OpamFilename.extract_job f local_dirname @@+ function
     | None -> Done (Up_to_date ())
-    | Some (Failure _) ->
-      Done (Not_available (Some "Could not extract archive",
-                           OpamFilename.(Base.to_string (basename f))^
-                            " (could not extract archive)"))
+    | Some (Failure s) ->
+      Done (Not_available (Some "Could not extract archive", s))
     | Some e -> Done (Not_available (None, Printexc.to_string e))
   in
   (match cache_dir with

